File: server-phoenix/lib/helios/daily_event_summary.ex
defmodule Helios.DailyEventSummary do
  use Ecto.Schema
  import Ecto.Query
  alias Helios.{Repo, DailyEventSummary}

  schema "daily_event_summaries" do
    field :source, :string, presence: true
    field :day, :date, presence: true
    field :count, :integer, presence: true
    timestamps([inserted_at: :created_at, type: :utc_datetime])
  end

  def archive!(event_scope) do
    event_scope
      |> group_by([e], e.source)
      |> group_by([e], fragment("day"))
      |> select([e], %{source: e.source, day: fragment("date(created_at) as day"), count: count(e.id)})
      |> Repo.all()
      |> Enum.each(fn (%{source: source, day: day, count: count}) -> Repo.insert(
      %DailyEventSummary {
        source: Atom.to_string(source),
        day: Date.from_iso8601!(day),
        count: count
      }) end)

    event_scope |> Repo.delete_all
  end
end

File: server-rails/app/models/daily_event_summary.rb
class DailyEventSummary < ApplicationRecord
  def self.archive!(event_scope)
    event_scope
      .group(:source)
      .group_by_day(:created_at, series: false)
      .count
      .each do |groups, count|
      DailyEventSummary.create!(
        source: groups[0],
        day: groups[1],
        count: count
      )
    end
    event_scope.delete_all
  end
end
